packages/deployer/subtasks/sync-proxy.js

```diff
@@ -0,0 +1,38 @@
+const { subtask } = require('hardhat/config');
+const logger = require('@synthetixio/core-js/utils/logger');
+const { initContractData } = require('../internal/process-contracts');
+const { ContractValidationError } = require('../internal/errors');
+const { SUBTASK_SYNC_PROXY } = require('../task-names');
+
+/**
+ * Compiles the Proxy's source, saves it to the deployments file, and checks
+ * that it does not have changes if it's already deployed.
+ */
+subtask(SUBTASK_SYNC_PROXY, 'Compile and sync the source from the Proxy.').setAction(
+  async (_, hre) => {
+    logger.subtitle('Syncing and compiling source from the Proxy');
+
+    const proxyName = hre.config.deployer.proxyContract;
+
+    await initContractData(proxyName);
```

Comment:
This is the main part of the fix, which initializes all the Proxy's contract data before running any validation.